src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/ReliableDictionaryCheckpointMananger.cs

```diff
@@ -0,0 +1,159 @@
+        {
+            ConditionalValue<IReliableDictionary<string, Checkpoint>> tryStore = await 
+                this.reliableStateManager.TryGetAsync<IReliableDictionary<string, Checkpoint>>(Constants.CheckpointDictionaryName);
+            EventProcessorEventSource.Current.Message("CheckpointStoreExistsAsync = {0}", tryStore.HasValue);
+            return tryStore.HasValue;
+        }
+
+        public async Task<bool> CreateCheckpointStoreIfNotExistsAsync(CancellationToken cancellationToken)
+        {
+            // Create or get access to the dictionary.
+            this.store = await reliableStateManager.GetOrAddAsync<IReliableDictionary<string, Dictionary<string, object>>>(Constants.CheckpointDictionaryName);
+            EventProcessorEventSource.Current.Message("CreateCheckpointStoreIfNotExistsAsync OK");
+            return true;
+        }
+
+        public async Task<Checkpoint> CreateCheckpointIfNotExistsAsync(string partitionId, CancellationToken cancellationToken)
+        {
+            Checkpoint existingCheckpoint = await GetWithRetry(partitionId, cancellationToken);
+
+            if (existingCheckpoint == null)
+            {
+                existingCheckpoint = new Checkpoint(1);
+                await PutWithRetry(partitionId, existingCheckpoint, cancellationToken);
+            }
+            EventProcessorEventSource.Current.Message("CreateCheckpointIfNotExists OK");
+
+            return existingCheckpoint;
+        }
+
+        public async Task<Checkpoint> GetCheckpointAsync(string partitionId, CancellationToken cancellationToken)
+        {
+            return await GetWithRetry(partitionId, cancellationToken);
+        }
+
+        public async Task UpdateCheckpointAsync(string partitionId, Checkpoint checkpoint, CancellationToken cancellationToken)
+        {
+            await PutWithRetry(partitionId, checkpoint, cancellationToken);
+        }
+
+        // Throws on error or if cancelled.
+        // Returns null if there is no entry for the given partition.
+        private async Task<Checkpoint> GetWithRetry(string partitionId, CancellationToken cancellationToken)
+        {
+            EventProcessorEventSource.Current.Message("Getting checkpoint for {0}", partitionId);
+
+            Checkpoint result = null;
+            Exception lastException = null;
+            for (int i = 0; i < Constants.RetryCount; i++)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                lastException = null;
+
+                try
+                {
+                    using (ITransaction tx = this.reliableStateManager.CreateTransaction())
+                    {
+                        ConditionalValue<Dictionary<string, object>> rawCheckpoint = await
+                            this.store.TryGetValueAsync(tx, partitionId, Constants.ReliableDictionaryTimeout, cancellationToken);
+
+                        await tx.CommitAsync();
```

Comment:
Do we need txn for get?

Reply:
Yes, the transaction is a required argment to the get method.